modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/NettyAllocator.java

```diff
@@ -353,4 +361,63 @@ public ByteBufAllocator getDelegate() {
             return delegate;
         }
     }
+
+    static class TrashingByteBuf extends WrappedByteBuf {
+
+        private volatile boolean trashed = false;
+
+        protected TrashingByteBuf(ByteBuf buf) {
+            super(buf);
+        }
+
+        @Override
+        public boolean release() {
+            if (refCnt() == 1) {
+                trashContent();
+            }
+            return super.release();
+        }
+
+        @Override
+        public boolean release(int decrement) {
+            if (refCnt() == decrement) {
+                trashContent();
+            }
+            return super.release(decrement);
+        }
+
+        private void trashContent() {
+            if (trashed == false) {
+                trashed = true;
+                for (var nioBuf : buf.nioBuffers()) {
+                    assert nioBuf.hasArray();
+                    var from = nioBuf.arrayOffset() + nioBuf.position();
+                    var to = from + nioBuf.remaining();
+                    Arrays.fill(nioBuf.array(), from, to, (byte) 0);
+                }
+            }
+        }
+    }
+
+    static class TrashingByteBufAllocator extends NoDirectBuffers {
```

Comment:
Oh wait I think we also need to wrap up any buffers created by `compositeBuffer()` and friends.

Reply:
I dont think we need. A composite is a collection of ByteBuf components. Each release will be delegated to the specific buffer that will trash bytes.

Reply:
You are right, added tests for composite buf and they dont pass. Will add a wrapper for composites too.